{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-27T18:47:46.164145Z",
     "start_time": "2020-08-27T18:47:46.040109Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mbam\u001b[m\u001b[m/               \u001b[34mfastq\u001b[m\u001b[m/             \u001b[34mpileup\u001b[m\u001b[m/\r\n",
      "data_manifest.rtf  \u001b[34mfilter\u001b[m\u001b[m/            \u001b[34mtable\u001b[m\u001b[m/\r\n"
     ]
    }
   ],
   "source": [
    "ls ../../testdata/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-27T18:47:49.681254Z",
     "start_time": "2020-08-27T18:47:49.557291Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "001.chr1.pileup                 002.csv\r\n",
      "001_A-B_A.chr1.filter1.HDR.csv  002.pileup\r\n",
      "001_A-B_A.chr1.filter1.pileup   002_A-B_A.chr1.filter1.HDR.csv\r\n",
      "002.bam                         002_A-B_A.chr1.filter1.pileup\r\n",
      "002.bam.bai                     003_A-B_A.chr1.filter1.HDR.csv\r\n",
      "002.chr1.pileup                 003_A-B_A.chr1.filter1.pileup\r\n"
     ]
    }
   ],
   "source": [
    "ls ../data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-27T18:48:19.181640Z",
     "start_time": "2020-08-27T18:48:19.134865Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneDetail</th>\n",
       "      <th>ExonicFunc</th>\n",
       "      <th>AAChange</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomAD_exome_ALL</th>\n",
       "      <th>avsnp150</th>\n",
       "      <th>SIFT_converted_rankscore</th>\n",
       "      <th>SIFT_pred</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>Polyphen2_HDIV_rankscore</th>\n",
       "      <th>Polyphen2_HDIV_pred</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>Polyphen2_HVAR_rankscore</th>\n",
       "      <th>Polyphen2_HVAR_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3632268</td>\n",
       "      <td>3632268</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>WRAP73</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>WRAP73:NM_017818:exon10:c.C993G:p.I331M</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0728</td>\n",
       "      <td>rs2760320</td>\n",
       "      <td>0.180</td>\n",
       "      <td>T</td>\n",
       "      <td>0.027</td>\n",
       "      <td>0.182</td>\n",
       "      <td>B</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.236</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4712657</td>\n",
       "      <td>4712657</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>AJAP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>AJAP1:NM_001042478:exon2:c.G787A:p.G263R,AJAP1...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5238</td>\n",
       "      <td>rs242056</td>\n",
       "      <td>0.165</td>\n",
       "      <td>T</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.764</td>\n",
       "      <td>D</td>\n",
       "      <td>0.892</td>\n",
       "      <td>0.615</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23417366</td>\n",
       "      <td>23417366</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>TCEA3</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>TCEA3:NM_003196:exon4:c.A263G:p.E88G</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>rs192746462</td>\n",
       "      <td>0.316</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.026</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.013</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39511892</td>\n",
       "      <td>39511892</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>BMP8A</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>BMP8A:NM_181809:exon3:c.G661A:p.E221K</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>rs764410592</td>\n",
       "      <td>0.427</td>\n",
       "      <td>D</td>\n",
       "      <td>0.923</td>\n",
       "      <td>0.492</td>\n",
       "      <td>P</td>\n",
       "      <td>0.811</td>\n",
       "      <td>0.568</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39769776</td>\n",
       "      <td>39769776</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>OXCT2</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>OXCT2:NM_022120:exon1:c.G1480A:p.D494N</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0379</td>\n",
       "      <td>rs150795467</td>\n",
       "      <td>0.477</td>\n",
       "      <td>D</td>\n",
       "      <td>0.988</td>\n",
       "      <td>0.604</td>\n",
       "      <td>D</td>\n",
       "      <td>0.784</td>\n",
       "      <td>0.556</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>chr9</td>\n",
       "      <td>136847266</td>\n",
       "      <td>136847266</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>C9orf172</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>C9orf172:NM_001080482:exon1:c.T2852G:p.M951R</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.784</td>\n",
       "      <td>D</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.764</td>\n",
       "      <td>D</td>\n",
       "      <td>0.994</td>\n",
       "      <td>0.807</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137232927</td>\n",
       "      <td>137232927</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SLC34A3</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>SLC34A3:NM_001177316:exon5:c.C448G:p.L150V,SLC...</td>\n",
       "      <td>...</td>\n",
       "      <td>1.303e-05</td>\n",
       "      <td>.</td>\n",
       "      <td>0.151</td>\n",
       "      <td>T</td>\n",
       "      <td>0.977</td>\n",
       "      <td>0.566</td>\n",
       "      <td>D</td>\n",
       "      <td>0.736</td>\n",
       "      <td>0.537</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>chrX</td>\n",
       "      <td>71122601</td>\n",
       "      <td>71122601</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>MED12</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>MED12:NM_005120:exon9:c.A1342C:p.T448P</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.501</td>\n",
       "      <td>D</td>\n",
       "      <td>0.979</td>\n",
       "      <td>0.899</td>\n",
       "      <td>D</td>\n",
       "      <td>0.865</td>\n",
       "      <td>0.807</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>chrX</td>\n",
       "      <td>108183752</td>\n",
       "      <td>108183752</td>\n",
       "      <td>-</td>\n",
       "      <td>CTT</td>\n",
       "      <td>exonic</td>\n",
       "      <td>COL4A6</td>\n",
       "      <td>.</td>\n",
       "      <td>nonframeshift insertion</td>\n",
       "      <td>COL4A6:NM_001287758:exon24:c.1971_1972insAAG:p...</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>rs146680910</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>chrX</td>\n",
       "      <td>141879335</td>\n",
       "      <td>141879335</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>MAGEC3</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>MAGEC3:NM_138702:exon3:c.A419T:p.D140V</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0753</td>\n",
       "      <td>rs12852593</td>\n",
       "      <td>0.784</td>\n",
       "      <td>D</td>\n",
       "      <td>0.969</td>\n",
       "      <td>0.548</td>\n",
       "      <td>D</td>\n",
       "      <td>0.599</td>\n",
       "      <td>0.489</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>177 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr      Start        End Ref  Alt    Func      Gene GeneDetail  \\\n",
       "0    chr1    3632268    3632268   G    C  exonic    WRAP73          .   \n",
       "1    chr1    4712657    4712657   G    A  exonic     AJAP1          .   \n",
       "2    chr1   23417366   23417366   T    C  exonic     TCEA3          .   \n",
       "3    chr1   39511892   39511892   G    A  exonic     BMP8A          .   \n",
       "4    chr1   39769776   39769776   C    T  exonic     OXCT2          .   \n",
       "..    ...        ...        ...  ..  ...     ...       ...        ...   \n",
       "172  chr9  136847266  136847266   T    G  exonic  C9orf172          .   \n",
       "173  chr9  137232927  137232927   C    G  exonic   SLC34A3          .   \n",
       "174  chrX   71122601   71122601   A    C  exonic     MED12          .   \n",
       "175  chrX  108183752  108183752   -  CTT  exonic    COL4A6          .   \n",
       "176  chrX  141879335  141879335   A    T  exonic    MAGEC3          .   \n",
       "\n",
       "                  ExonicFunc  \\\n",
       "0          nonsynonymous SNV   \n",
       "1          nonsynonymous SNV   \n",
       "2          nonsynonymous SNV   \n",
       "3          nonsynonymous SNV   \n",
       "4          nonsynonymous SNV   \n",
       "..                       ...   \n",
       "172        nonsynonymous SNV   \n",
       "173        nonsynonymous SNV   \n",
       "174        nonsynonymous SNV   \n",
       "175  nonframeshift insertion   \n",
       "176        nonsynonymous SNV   \n",
       "\n",
       "                                              AAChange  ... gnomAD_exome_ALL  \\\n",
       "0              WRAP73:NM_017818:exon10:c.C993G:p.I331M  ...           0.0728   \n",
       "1    AJAP1:NM_001042478:exon2:c.G787A:p.G263R,AJAP1...  ...           0.5238   \n",
       "2                 TCEA3:NM_003196:exon4:c.A263G:p.E88G  ...           0.0046   \n",
       "3                BMP8A:NM_181809:exon3:c.G661A:p.E221K  ...           0.0040   \n",
       "4               OXCT2:NM_022120:exon1:c.G1480A:p.D494N  ...           0.0379   \n",
       "..                                                 ...  ...              ...   \n",
       "172       C9orf172:NM_001080482:exon1:c.T2852G:p.M951R  ...                .   \n",
       "173  SLC34A3:NM_001177316:exon5:c.C448G:p.L150V,SLC...  ...        1.303e-05   \n",
       "174             MED12:NM_005120:exon9:c.A1342C:p.T448P  ...                .   \n",
       "175  COL4A6:NM_001287758:exon24:c.1971_1972insAAG:p...  ...                .   \n",
       "176             MAGEC3:NM_138702:exon3:c.A419T:p.D140V  ...           0.0753   \n",
       "\n",
       "        avsnp150  SIFT_converted_rankscore SIFT_pred  Polyphen2_HDIV_score  \\\n",
       "0      rs2760320                     0.180         T                 0.027   \n",
       "1       rs242056                     0.165         T                 0.999   \n",
       "2    rs192746462                     0.316         T                   0.0   \n",
       "3    rs764410592                     0.427         D                 0.923   \n",
       "4    rs150795467                     0.477         D                 0.988   \n",
       "..           ...                       ...       ...                   ...   \n",
       "172            .                     0.784         D                 0.999   \n",
       "173            .                     0.151         T                 0.977   \n",
       "174            .                     0.501         D                 0.979   \n",
       "175  rs146680910                         .         .                     .   \n",
       "176   rs12852593                     0.784         D                 0.969   \n",
       "\n",
       "     Polyphen2_HDIV_rankscore  Polyphen2_HDIV_pred  Polyphen2_HVAR_score  \\\n",
       "0                       0.182                    B                 0.048   \n",
       "1                       0.764                    D                 0.892   \n",
       "2                       0.026                    B                   0.0   \n",
       "3                       0.492                    P                 0.811   \n",
       "4                       0.604                    D                 0.784   \n",
       "..                        ...                  ...                   ...   \n",
       "172                     0.764                    D                 0.994   \n",
       "173                     0.566                    D                 0.736   \n",
       "174                     0.899                    D                 0.865   \n",
       "175                         .                    .                     .   \n",
       "176                     0.548                    D                 0.599   \n",
       "\n",
       "     Polyphen2_HVAR_rankscore  Polyphen2_HVAR_pred  \n",
       "0                       0.236                    B  \n",
       "1                       0.615                    P  \n",
       "2                       0.013                    B  \n",
       "3                       0.568                    P  \n",
       "4                       0.556                    P  \n",
       "..                        ...                  ...  \n",
       "172                     0.807                    D  \n",
       "173                     0.537                    P  \n",
       "174                     0.807                    P  \n",
       "175                         .                    .  \n",
       "176                     0.489                    P  \n",
       "\n",
       "[177 rows x 61 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('../data/002.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-27T19:00:05.363126Z",
     "start_time": "2020-08-27T18:57:00.621306Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;30;1mImporting ../data/002.csv for HDR detection\u001b[0m\n",
      "\u001b[1;35;2m20:57:00\u001b[0m : \u001b[1;30;1mStarting HDR analysis of ../data/002.csv on chr1 using 8 cores. [MIN_SIM=0.5, PAD=100]\u001b[0m\n",
      "\u001b[1;35;2m20:57:00\u001b[0m : \u001b[1;30;1mUsing 4 cores for hotspot detection on chr1\u001b[0m\n",
      "\u001b[1;35;2m20:57:31\u001b[0m : \u001b[1;30;1mDetected 32 putative HDR lanes in /Users/martinscience/Dropbox/Icke/Work/somVar/testdata/bam/003_A.chr1.bam.\u001b[0m\n",
      "     Chr        Pos Ref  Depth Alt  AltSum  AltRatio  totalAlt\n",
      "0   chr1    3632268   G    409   C     208  0.508557       208\n",
      "1   chr1    4712657   G    242   A     116  0.479339       116\n",
      "2   chr1   23417366   t    206   C     107  0.519417       107\n",
      "3   chr1   39511892   G    205   A      41  0.200000        41\n",
      "4   chr1   39511903   C    166   T     115  0.692771       115\n",
      "5   chr1   39769776   C    467   T     116  0.248394       116\n",
      "6   chr1   39769817   T    507   C      96  0.189349        96\n",
      "7   chr1  109192794   A    230   C     110  0.478261       110\n",
      "8   chr1  109192882   T    130   C      59  0.453846        59\n",
      "9   chr1  117118132   C    288   T     131  0.454861       131\n",
      "0   chr1  146135471   T    420   C     120  0.285714       120\n",
      "1   chr1  146164829   G    194   A      44  0.226804        44\n",
      "2   chr1  152155710   A   1039   T       4  0.003850         9\n",
      "3   chr1  152155713   G   1094   T       3  0.002742         5\n",
      "4   chr1  152155716   G   1045   T       2  0.001914         5\n",
      "5   chr1  152155717   G   1086   T       3  0.002762         7\n",
      "6   chr1  152155756   T    908   C      23  0.025330        31\n",
      "7   chr1  152155775   G   1072   T      15  0.013993        15\n",
      "8   chr1  152155782   A   1127   T       2  0.001775         8\n",
      "9   chr1  152155800   T   1184   -       1  0.000845         5\n",
      "10  chr1  152155817   C   1129   T       4  0.003543         7\n",
      "11  chr1  152155818   G   1098   T       6  0.005464         6\n",
      "12  chr1  152155824   A   1141   T       2  0.001753         8\n",
      "13  chr1  152155829   C   1042   T       6  0.005758        11\n",
      "14  chr1  152155841   G   1032   T      23  0.022287        23\n",
      "15  chr1  159815580   T    306   C     136  0.444444       136\n",
      "16  chr1  159815635   A    216   G      96  0.444444        96\n",
      "17  chr1  197061891   C    230   T      89  0.386957        89\n",
      "18  chr1  223394461   G    405   A     175  0.432099       175\n",
      "19  chr1  230868132   T     76   G       3  0.039474         5\n",
      "20  chr1  230868166   A    171   C      10  0.058479        10\n",
      "21  chr1  230868177   T    195   A       8  0.041026         8\n",
      "\u001b[1;35;2m20:57:31\u001b[0m : \u001b[1;30;1mUsing 8 cores for filtering HDR-rich mutations\u001b[0m\n",
      "\u001b[1;35;2m20:57:32\u001b[0m : \u001b[1;30;1mFound a total of 14 HDR-rich mutations\u001b[0m\n",
      "\u001b[1;35;2m20:57:32\u001b[0m : \u001b[1;30;1mUsing 8 cores for HDR computation\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Gene</th>\n",
       "      <th>HDRcand</th>\n",
       "      <th>HDRcount</th>\n",
       "      <th>HDRinfo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3632268</td>\n",
       "      <td>3632268</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>WRAP73</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4712657</td>\n",
       "      <td>4712657</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>AJAP1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>23417366</td>\n",
       "      <td>23417366</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>TCEA3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39511892</td>\n",
       "      <td>39511892</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BMP8A</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>no similarity in HDR-pattern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39769776</td>\n",
       "      <td>39769776</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OXCT2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>∆41&lt;Ref:99%(532)&gt;&lt;Alt:99%(179)&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>109192794</td>\n",
       "      <td>109192794</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>KIAA1324</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>∆88&lt;Ref:96%(81)&gt;&lt;Alt:90%(89)&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>117118132</td>\n",
       "      <td>117118132</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>TRIM45</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146135471</td>\n",
       "      <td>146135471</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NBPF10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>146164829</td>\n",
       "      <td>146164829</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NOTCH2NL</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>152155775</td>\n",
       "      <td>152155775</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>RPTN</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>no similarity in HDR-pattern</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>159815635</td>\n",
       "      <td>159815635</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>FCRL6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>∆-55&lt;Ref:97%(227)&gt;&lt;Alt:99%(141)&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>197061891</td>\n",
       "      <td>197061891</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>F13B</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>223394461</td>\n",
       "      <td>223394461</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>CCDC185</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no HDR in vincinity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>230868212</td>\n",
       "      <td>230868212</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C1orf198</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>∆-46&lt;Ref:96%(270)&gt;&lt;Alt:71%(7)&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chr      Start        End Ref Alt      Gene  HDRcand  HDRcount  \\\n",
       "0   chr1    3632268    3632268   G   C    WRAP73        1         0   \n",
       "1   chr1    4712657    4712657   G   A     AJAP1        1         0   \n",
       "2   chr1   23417366   23417366   T   C     TCEA3        1         0   \n",
       "3   chr1   39511892   39511892   G   A     BMP8A        2         0   \n",
       "4   chr1   39769776   39769776   C   T     OXCT2        2         1   \n",
       "5   chr1  109192794  109192794   A   C  KIAA1324        2         1   \n",
       "6   chr1  117118132  117118132   C   T    TRIM45        1         0   \n",
       "7   chr1  146135471  146135471   T   C    NBPF10        1         0   \n",
       "8   chr1  146164829  146164829   G   A  NOTCH2NL        1         0   \n",
       "9   chr1  152155775  152155775   G   T      RPTN       13         0   \n",
       "10  chr1  159815635  159815635   A   G     FCRL6        2         1   \n",
       "11  chr1  197061891  197061891   C   T      F13B        1         0   \n",
       "12  chr1  223394461  223394461   G   A   CCDC185        1         0   \n",
       "13  chr1  230868212  230868212   T   G  C1orf198        3         1   \n",
       "\n",
       "                             HDRinfo  \n",
       "0                no HDR in vincinity  \n",
       "1                no HDR in vincinity  \n",
       "2                no HDR in vincinity  \n",
       "3       no similarity in HDR-pattern  \n",
       "4    ∆41<Ref:99%(532)><Alt:99%(179)>  \n",
       "5      ∆88<Ref:96%(81)><Alt:90%(89)>  \n",
       "6                no HDR in vincinity  \n",
       "7                no HDR in vincinity  \n",
       "8                no HDR in vincinity  \n",
       "9       no similarity in HDR-pattern  \n",
       "10  ∆-55<Ref:97%(227)><Alt:99%(141)>  \n",
       "11               no HDR in vincinity  \n",
       "12               no HDR in vincinity  \n",
       "13    ∆-46<Ref:96%(270)><Alt:71%(7)>  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append('../code')\n",
    "from io import StringIO\n",
    "from subprocess import Popen, PIPE, run\n",
    "from HDR_run import HDR_master\n",
    "\n",
    "home = '/Users/mahtin'\n",
    "home = '/Users/martinscience'\n",
    "somvar = f'{home}/Dropbox/Icke/Work/somVar'\n",
    "data_path = f'{somvar}/testdata'\n",
    "static_path = f'{home}/Dropbox/Icke/Work/static'\n",
    "hdr_path = f'{somvar}/HDRtest'\n",
    "bam_file = f'{data_path}/bam/003_A.chr1.bam'\n",
    "pileup_file = f'{data_path}/pileup/003_A.chr1.pileup'\n",
    "mut_file = '../data/002.csv'\n",
    "\n",
    "chrom = 'chr1'\n",
    "genome_split_path = f'{static_path}/genome/gatk/hg38/split'\n",
    "# chrom = ''\n",
    "\n",
    "threads = 8\n",
    "HDR_config = {\n",
    "    \"minAltSum\": 5,\n",
    "    \"minAltRatio\": 0.1,\n",
    "    \"maxAltRatio\": 1,\n",
    "    \"MINQ\": 15,\n",
    "    \"MINSIM\": .50,\n",
    "    \"PAD\": 100,\n",
    "    \"MINq\": 10,\n",
    "    \"MinAltSupport\": 5,\n",
    "    \"MinHDRCount\": 1,\n",
    "    \"pile2hotspot\": '../shell/pile2hotspot.mawk',#\n",
    "    \"pile2hotspot_chrom\": '../shell/pile2hotspot_chrom.mawk',\n",
    "    \"editbam\": '../shell/editbam.mawk',\n",
    "    \"bam2csv\": '../shell/bam2csv.mawk',\n",
    "    \"genome_split_path\": genome_split_path # the path to the folder with chrom-split genomes (chr1.fa..)\n",
    "}\n",
    "HDR_master(mut_file, bam_file=bam_file, chrom=chrom, threads=threads, HDR_config=HDR_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-23T14:07:32.114469Z",
     "start_time": "2020-08-23T14:07:31.983073Z"
    }
   },
   "outputs": [],
   "source": [
    "ls data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-23T19:25:07.814840Z",
     "start_time": "2020-08-23T19:25:07.772067Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "home = '/Users/mahtin'\n",
    "path = '/Users/mahtin/Dropbox/Icke/Work/somVar/HDRtest'\n",
    "bam_file = f'{path}/data/002.bam '\n",
    "chrom = 'chr1'\n",
    "# chrom = ''\n",
    "pileup_file = f'{path}/data/002.pileup'\n",
    "threads = 8\n",
    "mut_file = f'{path}/data/002.csv'\n",
    "mut_df = pd.read_csv(mut_file, sep='\\t').loc[:, [\n",
    "        'Chr', 'Start', 'End', 'Ref', 'Alt', 'Gene']]\n",
    "# make Chr column categorical for sorting .. and sort\n",
    "chrom_list = [f\"chr{i}\" for i in range(23)] + ['chrX', 'chrY']\n",
    "mut_df['Chr'] = pd.Categorical(mut_df['Chr'], chrom_list)\n",
    "mut_df = mut_df.sort_values(['Chr', 'Start'])\n",
    "mut_df = mut_df.query('Chr == @chrom')\n",
    "mut_df\n",
    "mut_split = np.array_split(mut_df, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-23T19:14:23.270088Z",
     "start_time": "2020-08-23T19:14:23.267394Z"
    }
   },
   "outputs": [],
   "source": [
    "mut_df = mut_split[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-23T19:30:19.182347Z",
     "start_time": "2020-08-23T19:30:19.170669Z"
    }
   },
   "outputs": [],
   "source": [
    "anno_df = mut_df.sort_values(['Chr', 'Start']).iloc[0:,:5]\n",
    "anno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-23T19:38:46.184923Z",
     "start_time": "2020-08-23T19:38:46.179603Z"
    }
   },
   "outputs": [],
   "source": [
    "def reduce_regions(df, padding):\n",
    "    '''\n",
    "    takes a mutation list and returns a region list using padding\n",
    "    overlapping regions are reduced to one using the gap strategy\n",
    "    '''\n",
    "\n",
    "    df = df.sort_values('Start')\n",
    "    df['Start'] = df['Start'] - padding\n",
    "    df['End'] = df['End'] + padding\n",
    "    # find the break points\n",
    "    # if Start is greater than previous End (using shift), this is a gap --> df['gap'] = 1\n",
    "    df['gap'] = df['Start'].gt(df['End'].shift()).astype('int')\n",
    "    # id different reads according to gap\n",
    "    # cumulative sum does not increase at df['gap'] == 0 and so these consecutive stretches are grouped together\n",
    "    df['gap'] = df['gap'].cumsum()\n",
    "    # groupby the coverage break group and condense individual coverage islands\n",
    "    # agg has to contain the neccessary shared columns TransLength because it is needed for coverage computation\n",
    "    df = df.groupby('gap').agg({'Chr':'first','Start': 'min', 'End':'max'})\n",
    "    return df.reset_index('gap').drop(columns='gap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-23T19:37:26.582555Z",
     "start_time": "2020-08-23T19:37:26.547975Z"
    }
   },
   "outputs": [],
   "source": [
    "df = anno_df\n",
    "anno_df\n",
    "padding = 100\n",
    "df['Start'] = df['Start'] - padding\n",
    "df['End'] = df['End'] + padding\n",
    "df['gap'] = df['Start'].gt(df['End'].shift()).astype('int')\n",
    "df['gap'] = df['gap'].cumsum()\n",
    "df = df.groupby('gap').agg({'Chr':'first','Start': 'min', 'End':'max'}).reset_index().drop(columns='gap')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-08-23T19:38:49.099313Z",
     "start_time": "2020-08-23T19:38:49.076795Z"
    }
   },
   "outputs": [],
   "source": [
    "bed_df = reduce_regions(anno_df, 100)\n",
    "bed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(columns=[\n",
    "        'Chr',\n",
    "        'Start',\n",
    "        'End',\n",
    "        'Ref',\n",
    "        'Alt',\n",
    "        'Gene',\n",
    "        'HDRcand',\n",
    "        'HDRcount',\n",
    "        'HDRinfo',\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
